src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -0,0 +1,630 @@
+            checksum += WriteLeftAlignedBytesAndGetChecksum(GnuVersionBytes, buffer.Slice(FieldLocations.Version, FieldLengths.Version));
+            return checksum;
+        }
+
+        // Writes the posix fields shared by ustar, pax and gnu, into the specified spans.
+        private int WritePosixAndGnuSharedFields(Span<byte> buffer)
+        {
+            int checksum = 0;
+
+            if (!string.IsNullOrEmpty(_uName))
+            {
+                checksum += WriteAsAsciiString(_uName, buffer, FieldLocations.UName, FieldLengths.UName);
+            }
+
+            if (!string.IsNullOrEmpty(_gName))
+            {
+                checksum += WriteAsAsciiString(_gName, buffer, FieldLocations.GName, FieldLengths.GName);
+            }
+
+            if (_devMajor > 0)
+            {
+                checksum += WriteAsOctal(_devMajor, buffer, FieldLocations.DevMajor, FieldLengths.DevMajor);
+            }
+
+            if (_devMinor > 0)
+            {
+                checksum += WriteAsOctal(_devMinor, buffer, FieldLocations.DevMinor, FieldLengths.DevMinor);
+            }
+
+            return checksum;
+        }
+
+        // Saves the gnu-specific fields into the specified spans.
+        private int WriteGnuFields(Span<byte> buffer)
+        {
+            int checksum = WriteAsTimestamp(_aTime, buffer, FieldLocations.ATime, FieldLengths.ATime);
+            checksum += WriteAsTimestamp(_cTime, buffer, FieldLocations.CTime, FieldLengths.CTime);
+
+            if (_gnuUnusedBytes != null)
+            {
+                checksum += WriteLeftAlignedBytesAndGetChecksum(_gnuUnusedBytes, buffer.Slice(FieldLocations.GnuUnused, FieldLengths.AllGnuUnused));
+            }
+
+            return checksum;
+        }
+
+        // Writes the current header's data stream into the archive stream.
+        private static void WriteData(Stream archiveStream, Stream dataStream, long actualLength)
+        {
+            dataStream.CopyTo(archiveStream); // The data gets copied from the current position
+            int paddingAfterData = TarHelpers.CalculatePadding(actualLength);
+            archiveStream.Write(new byte[paddingAfterData]);
+        }
+
+        // Dumps into the archive stream an extended attribute entry containing metadata of the entry it precedes.
+        private static Stream? GenerateExtendedAttributesDataStream(IEnumerable<KeyValuePair<string, string>> extendedAttributes)
+        {
+            MemoryStream? dataStream = null;
+            foreach ((string attribute, string value) in extendedAttributes)
+            {
+                // Need to do this because IEnumerable has no Count property
+                dataStream ??= new MemoryStream();
+
+                byte[] entryBytes = GenerateExtendedAttributeKeyValuePairAsByteArray(Encoding.UTF8.GetBytes(attribute), Encoding.UTF8.GetBytes(value));
+                dataStream.Write(entryBytes);
```

Comment:
We shouldn't need to allocate byte[]s for attribute and value in order to do this.